File: rust-lab/src/mergesort.rs
// Author : Thomas Minier
// Apply Merge sort on two slices
fn merge(left : &[i32], right : &[i32]) -> Vec<i32> {
    let mut res = vec!();
    let (mut left_set, mut right_set) = (left, right);

    while (!left_set.is_empty()) && (!right_set.is_empty()) {
        let (left_head, left_tail) = left_set.split_first().unwrap();
        let (right_head, right_tail) = right_set.split_first().unwrap();
        if left_head <= right_head {
            res.push(*left_head);
            left_set = left_tail;
        } else {
            res.push(*right_head);
            right_set = right_tail;
        }
    }
    // drain collections
    for x in left_set {
        res.push(*x);
    }
    for x in right_set {
        res.push(*x);
    }
    res
}

// Sort a slice using the Merge Sort algorithm
pub fn merge_sort(list : &[i32]) -> Vec<i32> {
    if list.len() <= 1 {
        return list.to_vec();
    }
    let (left, right) = list.split_at(list.len() / 2);
    return merge(&merge_sort(left), &merge_sort(right));
}

File: haskell/mergesort.hs
-- Implementation of the merge sort algorithm in Haskell
-- Author : Thomas Minier

merge :: (Ord a) => [a] -> [a] -> [a]
merge x [] = x
merge [] y = y
merge left@(x:tx) right@(y:ty) = if x <= y then x:(merge tx right) else y:(merge left ty)

halves :: (Ord a) => [a] -> ([a], [a])
halves x =
  let n = length x `div` 2
  in ((take n x), (drop n x))

mergesort :: (Ord a) => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort x =
  let (left, right) = halves x
  in merge (mergesort left) (mergesort right)
